<launch>

    <!-- defines name of the namespace of the drone -->
    <arg name="UAV_NAME" default="$(optenv UAV_NAME)" />
    <arg name="RUN_TYPE" default="simulation" />

    <arg name="custom_config" default="" />

    <!-- will it run using GNU debugger? -->
    <arg name="DEBUG" default="false" />
    <arg unless="$(arg DEBUG)" name="launch_prefix_debug" value="" />
    <arg if="$(arg DEBUG)" name="launch_prefix_debug" value="debug_roslaunch" />

    <!-- will it run as standalone nodelet or using a nodelet manager? -->
    <arg name="standalone" default="true" />
    <arg name="manager" default="$(arg UAV_NAME)_waypointflier_manager" />
    <arg name="n_threads" default="8" />
    <arg unless="$(arg standalone)" name="nodelet" value="load" />
    <arg if="$(arg standalone)" name="nodelet" value="standalone" />
    <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg manager)" />
    <arg if="$(arg standalone)" name="nodelet_manager" value="" />

    <!-- Namespace - all topics, services and parameters will be remapped using this namespace as a prefix (eg.: "/waypoint_flier/odom_uav_in" to "/uav1/waypoint_flier/odom_uav_in") -->
    <group ns="$(arg UAV_NAME)">

        <!-- FormationPlanning nodelet -->
        <node pkg="nodelet" type="nodelet" name="rbl_controller"
            args="$(arg nodelet) formation_control/RBLController $(arg nodelet_manager)"
            launch-prefix="$(arg launch_prefix_debug)" output="screen">

        <!-- lidar clustering to measure distance to leader -->
        <!-- <node name="lidar_clusters$(arg UAV_NAME)" pkg="rbl_controller" type="lidar_clusters" output="screen"/> -->
            <!-- Parameters loaded from launch file -->
            <param name="simulation" value="$(eval arg('RUN_TYPE') == 'simulation')" />
            <param name="uav_name" value="$(arg UAV_NAME)" />
            <!-- TODO: move this to config .yaml so you can change if simulation or experiment -->
              
            <!-- <param name="control_frame" value="$(arg UAV_NAME)/hector_origin" /> -->
            
           <!-- <param name="control_frame" value="$(arg UAV_NAME)/fcu" /> -->


            <!-- <param name="control_frame" value="$(arg UAV_NAME)/point_lio_origin" />  -->
            <param name="control_frame" value="$(arg UAV_NAME)/world_origin" />


            <!-- <param name="control_frame" value="$(arg UAV_NAME)/livox" /> -->
            <!-- <node pkd="tf2_ros" type="static_transform_publisher" name="point_lio_init_to_control_frame" args="0 0 0 0 0.785 0 $(arg UAV_NAME)/point_lio_init point_lio_init_for_control_frame" />
            <param name="control_frame" value="$(arg UAV_NAME)/point_lio_init_for_control_frame" /> -->
            
              <!-- ROS parameters config file -->
            <rosparam file="$(find rbl_controller)/config/default.yaml" />
            <rosparam file="$(arg custom_config)" />
            <!-- <rosparam file="$(find formation_controller_susd)/config/$(arg RUN_TYPE).yaml" /> -->

            <!-- Subscribers (note that the topic names will also be remapped using the namespace as a prefix) -->
            <remap from="~tracker_cmd_in" to="control_manager/tracker_cmd" />

            <!-- Service clients (note that the service names will also be remapped using the namespace as a prefix) -->
            <remap from="~ref_pos_out" to="control_manager/reference" />
            <remap from="~goto_out" to="control_manager/goto" />
            <!-- Service servers (note that the service names will also be remapped using the namespace as a prefix) -->
            <remap from="~control_activation_in" to="~activation" />
            <remap from="~control_deactivation_in" to="~deactivation" />
            <remap from="~fly_to_start_in" to="~fly_to_start" />
            <remap from="~save_to_csv_in" to="~save_to_csv" />

            <!-- Publishers (note that the service names will also be remapped using the namespace as a prefix) -->
            <remap from="~destination_out" to="~destination_vis" />
            <remap from="~centroid_out" to="~centroid_vis" />
            <remap from="~position_out" to="~position_vis" />
            <remap from="~obstacle_markers_out" to="~obstacles_vis" />
            <remap from="~neighbors_markers_out" to="~neighbors_vis" />
            <remap from="~hull_markers_out" to="~hull_vis" />

            

            

        </node>

    </group>

</launch>
